### PR TITLE
Asynchronous Uploads

### PR DESCRIPTION
**Pull request by sumnerevans**
Implements the recently merged MSC2246 Asynchronous Uploads for Synapse's media repository.

Closes #12484

* [x] Configuration options rate limiting media creation
* [x] Configuration options for how long to wait until media IDs expire
* [x] `POST /_matrix/media/v1/create`
* [x] `PUT /_matrix/media/v3/upload/{serverName}/{mediaId}`
* [x] `timeout_ms` support on `/download`
* [x] `timeout_ms` support on `/thumbnail`
* [x] Option to restrict how many open media uploads a given user can have.

Signed-off-by: Sumner Evans <sumner@beeper.com>

---

Keywords: async

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by clokep** [comment]
@sumnerevans I'm going to remove the review request from this since you marked it as a draft; if you *are* still looking for a review please shout about what bits you're looking for a review of (or why it is in draft but you want a review). Thanks! 👍

**Comment by sumnerevans** [comment]
@clokep I accidentally opened as a non-draft PR. And then the automatic reviewer adder functionality kicked in and I could not remove it. Sorry about that!

**Comment by clokep** [comment]
Not an issue! If it doesn't kick in when un-drafting the PR just shout! 👍

**Comment by sumnerevans** [comment]
Blocked by https://github.com/matrix-org/complement/pull/638

File: synapse/media/media_repository.py [hunk 1]
```diff
@@ -174,6 +178,108 @@ def mark_recently_accessed(self, server_name: Optional[str], media_id: str) -> N
...
+                403,
+                "Only the creator of the media ID can upload to it",
+                errcode=Codes.FORBIDDEN,
+            )
+
+        if media.get("media_length") is not None:
+            raise SynapseError(
+                409,
+                "Media ID already has content",
+                errcode="M_CANNOT_OVERWRITE_MEDIA",
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:245]: We can add this constant to `synapse/api/errors.py` (replace all usage)

File: synapse/media/media_repository.py [hunk 2]
```diff
@@ -174,6 +178,108 @@ def mark_recently_accessed(self, server_name: Optional[str], media_id: str) -> N
...
+        """Update the content of the given media ID.
+        Args:
+            media_id: The media ID to replace.
+            media_type: The content type of the file.
+            upload_name: The name of the file, if provided.
+            content: A file like object that is the content to store
+            content_length: The length of the content
+            auth_user: The user_id of the uploader
+        """
+        file_info = FileInfo(server_name=None, file_id=media_id)
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:269]: I see we do the [same `server_name=None` for `create_content(...)`](https://github.com/matrix-org/synapse/blob/d93912042191d30ff1f7aa41d9f0779a609caca8/synapse/media/media_repository.py#L200) but I don't see the nuance on why we don't specify the server name?

File: synapse/media/media_repository.py [hunk 3]
```diff
@@ -216,8 +322,62 @@ async def create_content(
...
+                media to be uploaded.
+        Returns:
+            Either the info dictionary for the given local media ID or
+            ``None``. If ``None``, then no further processing is necessary as
+            this function will send the necessary JSON response.
+        """
+        wait_until = self.clock.time_msec() + max_timeout_ms
+        while True:
+            # Get the info for the media
+            media_info = await self.store.get_local_media(media_id)
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:353]: This isn't the fault of this PR but all of these `media_info` and `media` dicts are pretty hard to follow what's available. Especially trying to follow `unused_expires_at` and `quarantined_by`

File: synapse/media/media_repository.py [hunk 4]
```diff
@@ -292,27 +456,31 @@ async def get_remote_media(
...
-        if responder:
+        if responder and media_info:
             media_type = media_info["media_type"]
             media_length = media_info["media_length"]
             upload_name = name if name else media_info["upload_name"]
             await respond_with_responder(
                 request, responder, media_type, media_length, upload_name
             )
         else:
-            respond_404(request)
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:307]: Should we still respond with a normal `404` in some cases?

File: synapse/rest/media/create_resource.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        (
+            reached_pending_limit,
+            first_expiration_ts,
+        ) = await self.media_repo.reached_pending_media_limit(
+            requester.user, self.max_pending_media_uploads
+        )
+        if reached_pending_limit:
+            raise LimitExceededError(
+                msg="You have too many uploads pending. Please finish uploading your "
+                "existing files.",
```
Comment (MadLittleMods) [inline synapse/rest/media/create_resource.py:74]: Is there any way to cancel a pending upload? I didn't see anything in the MSC

File: synapse/rest/media/thumbnail_resource.py [hunk 6]
```diff
@@ -142,8 +152,13 @@ async def _select_or_generate_local_thumbnail(
...
         desired_method: str,
         desired_type: str,
+        max_timeout_ms: int,
     ) -> None:
-        media_info = await self.store.get_local_media(media_id)
+        media_info = await self.media_repo.get_local_media_info(
+            request, media_id, max_timeout_ms
+        )
+        if not media_info:
+            return
```
Comment (MadLittleMods) [inline synapse/rest/media/thumbnail_resource.py:161]: This condition is duplicated below. Should we use the 404 branch below instead?

File: synapse/rest/media/thumbnail_resource.py [hunk 7]
```diff
@@ -208,8 +223,13 @@ async def _select_or_generate_remote_thumbnail(
...
         desired_method: str,
         desired_type: str,
+        max_timeout_ms: int,
     ) -> None:
-        media_info = await self.media_repo.get_remote_media_info(server_name, media_id)
+        media_info = await self.media_repo.get_remote_media_info(
+            server_name, media_id, max_timeout_ms
+        )
+        if not media_info:
+            return
```
Comment (MadLittleMods) [inline synapse/rest/media/thumbnail_resource.py:232]: Related to the `_select_or_generate_local_thumbnail` behavior mentioned above

```suggestion
            respond_404(request)
            return
```

Also applies to `_respond_remote_thumbnail` below

File: synapse/storage/databases/main/media_repository.py [hunk 8]
```diff
@@ -361,6 +406,30 @@ async def mark_local_media_as_safe(self, media_id: str, safe: bool = True) -> No
...
+            A tuple of two integers: the total pending media requests and the earliest
+            expiration timestamp.
+        """
+
+        def get_pending_media_txn(txn: LoggingTransaction) -> Tuple[int, int]:
+            sql = (
+                "SELECT COUNT(*), MIN(unused_expires_at)"
+                " FROM local_media_repository"
+                " WHERE user_id = ?"
+                " AND quarantined_by IS NULL"
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/media_repository.py:421]: Why do we care about excluding media that is quarantined?

This doesn't appear to be some efficient use of an index thing to get a smaller dataset to work with.

File: tests/media/test_media_storage.py [hunk 9]
```diff
@@ -294,7 +294,9 @@ def _req(
         self.assertEqual(
```
Comment (MadLittleMods) [inline tests/media/test_media_storage.py:294]: Are there tests for all this new behavior somewhere? I don't see Complement tests

File: synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Add new colums to the `local_media_repository` to keep track of when the
+-- media ID must be used by. This is to support async uploads (see MSC2246).
+
+ALTER TABLE local_media_repository
+  ADD COLUMN unused_expires_at BIGINT DEFAULT NULL;
```
Comment (MadLittleMods) [inline synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql:20]: We might need to add an index here to make it fast like we had to to do for `/timestamp_to_event`, https://github.com/matrix-org/synapse/pull/14799/files

File: synapse/media/media_repository.py [hunk 11]
```diff
@@ -216,8 +322,62 @@ async def create_content(
...
+            media_info = await self.store.get_local_media(media_id)
+            if not media_info:
+                respond_404(request)
+                return None
+
+            if media_info["quarantined_by"]:
+                logger.info("Media is quarantined")
+                respond_404(request)
+                return None
+
```
Comment (MadLittleMods) [inline synapse/media/media_repository.py:362]: I think we could add an extra case to check that `unused_expires_at` isn't already past the expiration

File: synapse/rest/media/create_resource.py (hunk 5 above)
Comment (sumnerevans) [reply synapse/rest/media/create_resource.py:74]: No, there is not. It will just expire after the configured timeout.

File: synapse/media/media_repository.py (hunk 2 above)
Comment (sumnerevans) [reply synapse/media/media_repository.py:269]: I'm not sure why, either, but according to the [comment on `server_name` in the `FileInfo` type](https://github.com/matrix-org/synapse/blob/d93912042191d30ff1f7aa41d9f0779a609caca8/synapse/media/_base.py#L345), it's supposed to be `None` for local.

I assume that there's somewhere in the code that relies on it being `None` (maybe remote media cleanup?) but I'm not sure. Regardless I don't think that it's in scope of this PR to fix this quirk.

File: synapse/storage/databases/main/media_repository.py (hunk 8 above)
Comment (sumnerevans) [reply synapse/storage/databases/main/media_repository.py:421]: I guess it really doesn't matter to exclude quarantined here. Probably better if we don't, since if someone has media quarantined, they probably are a malicious party and shouldn't be allowed to upload more media.

File: tests/media/test_media_storage.py (hunk 9 above)
Comment (sumnerevans) [reply tests/media/test_media_storage.py:294]: I got a start on it here: https://github.com/matrix-org/complement/pull/641

I'm not sure how to configure this PR to actually run the test, though.

File: synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql (hunk 10 above)
Comment (sumnerevans) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql:20]: Would I need to do it as a background task like `events_bg_updates` does?

File: tests/media/test_media_storage.py (hunk 9 above)
Comment (MadLittleMods) [reply tests/media/test_media_storage.py:294]: Same branch name is usually the solution but not sure how it works for forks

File: synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql (hunk 10 above)
Comment (MadLittleMods) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql:20]: Do we even need an index? (I was just making an educated guess based on previous experience)

> do it as a background task like events_bg_updates does?

:+1: Example https://github.com/matrix-org/synapse/pull/14799/files

**Comment by clokep** [comment]
@sumnerevans Hello! 👋 Wanted to check in on the state of this as I'd love for Synapse to support Matrix 1.7 soon (#15661). I see a few open threads and a couple of conflicts; are you hoping to get back to updating this soon?

**Comment by sumnerevans** [comment]
The biggest thing is figuring out how to deal with the complement tests. Since they have to be in a branch in Beeper's fork of complement, I don't know how to make this PR use those tests.

I can deal with the other two open threads on Friday.

**Comment by clokep** [comment]
> Since they have to be in a branch in Beeper's fork of complement, I don't know how to make this PR use those tests.

Looks like this is matrix-org/complement#641. I can run them manually (or I can re-push both branches to matrix-org organization & kick off a manual build to get the branch matching).

File: synapse/media/media_repository.py [hunk 12]
```diff
@@ -473,8 +473,12 @@ async def get_remote_media(
             await respond_with_responder(
                 request, responder, media_type, media_length, upload_name
             )
-        else:
+        elif media_info:
+            # The media exists, but the responder doesn't meaning that this file hasn't
```
Comment (jplatte) [inline synapse/media/media_repository.py:477]:
```suggestion
            # The media exists but the responder doesn't, meaning that this file hasn't
```

**Comment by clokep** [comment]
> > Since they have to be in a branch in Beeper's fork of complement, I don't know how to make this PR use those tests.
> 
> Looks like this is [matrix-org/complement#641](https://github.com/matrix-org/complement/pull/641). I can run them manually (or I can re-push both branches to matrix-org organization & kick off a manual build to get the branch matching).

I've now done this at https://github.com/matrix-org/synapse/actions/runs/5858699560

File: synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql (hunk 10 above)
Comment (sumnerevans) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: image

I can't create the index concurrently, which means that it will probably take forever to create this index and block startup, right?

Comment (clokep) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: We probably only need an index if we're frequently looking things up by `unused_expires_at`, I don't see any code that does this so we likely don't need one?

----

Regardless, the way to do this would be to create it in the background by doing something like this:

https://github.com/matrix-org/synapse/blob/2ad91ec628126753590c1a90c432270d6c8fa8fd/synapse/storage/databases/main/event_push_actions.py#L284-L290

And then adding a background job in the schema file, something like:

https://github.com/matrix-org/synapse/blob/666ae877292d4747b9441105e3df8558f7a335c0/synapse/storage/schema/main/delta/72/06thread_notifications.sql#L26-L27

Comment (sumnerevans) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: I'm using the `unused_expires_at` in a query here: https://github.com/matrix-org/synapse/pull/15503/files#diff-f05c2d0a285e2b60b93cbffe6283864c9b23c19984fe7e4511604797413acc06R424

Comment (clokep) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: But you're also selecting on `user_id`, which *does* have an index already. (The index is `user_id, created_ts`.)

Comment (sumnerevans) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: Is a linear scan on the user's entire list of media going to be fast though?

Comment (clokep) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: It depends on how much media the user has, I think. Would it make more sense to re-use that index and change all the conditions to be based on `created_ts` instead of a new expiration time? E.g.

```sql
SELECT COUNT(*), MIN(created_ts)
FROM local_media_repository
WHERE user_id = ?
	AND created_ts > ?
	AND media_length IS NULL
)
```

And then provide the arguments such that you do `self._clock.time_msec() - unused_expiration_time`.

Comment (sumnerevans) [reply synapse/storage/schema/main/delta/78/03_add_unused_expires_at_for_media.sql]: Updated this PR to do expirations based on `created_ts`. If in the future we want to have variable expirations, we will need to invent something smart here, but for now this is simple enough.

File: synapse/rest/media/upload_resource.py [hunk 13]
```diff
@@ -105,3 +115,44 @@ async def _async_render_POST(self, request: SynapseRequest) -> None:
...
+                    content,
+                    content_length,
+                    requester.user,
+                )
+            except SpamMediaException:
+                # For uploading of media we want to respond with a 400, instead of
+                # the default 404, as that would just be confusing.
+                raise SynapseError(400, "Bad content")
+
+            logger.info("Uploaded content to URI %r", media_id)
```
Comment (clokep) [inline synapse/rest/media/upload_resource.py:157]:
```suggestion
            logger.info("Uploaded content for media ID %r", media_id)
```

File: synapse/storage/databases/main/media_repository.py [hunk 14]
```diff
@@ -363,6 +408,35 @@ async def mark_local_media_as_safe(self, media_id: str, safe: bool = True) -> No
...
+        """
+
+        def get_pending_media_txn(txn: LoggingTransaction) -> Tuple[int, int]:
+            sql = (
+                "SELECT COUNT(*), MIN(created_at)"
+                " FROM local_media_repository"
+                " WHERE user_id = ?"
+                " AND created_at > ?"
+                " AND media_length IS NULL"
+            )
```
Comment (clokep) [inline synapse/storage/databases/main/media_repository.py:425]: I usually find these clearer as multi-line strings:

```suggestion
            sql = """
            SELECT COUNT(*), MIN(created_at)
            FROM local_media_repository
            WHERE user_id = ?
                AND created_at > ?
                AND media_length IS NULL
            """
```

File: synapse/storage/databases/main/media_repository.py [hunk 15]
```diff
@@ -363,6 +408,35 @@ async def mark_local_media_as_safe(self, media_id: str, safe: bool = True) -> No
...
+            txn.execute(
+                sql,
+                (
+                    user_id.to_string(),
+                    self._clock.time_msec() - self.unused_expiration_time,
+                ),
+            )
+            row = txn.fetchone()
+            if not row:
+                raise StoreError(404, "Failed to count pending media for user")
```
Comment (clokep) [inline synapse/storage/databases/main/media_repository.py:435]: If this is a 404 doesn't that mean there's no pending media?

```suggestion
            if not row:
                return 0, 0
```

**Review by clokep — COMMENTED** [review]
There's a couple of questions and a bunch of minor comments.

File: synapse/media/media_repository.py [hunk 16]
```diff
@@ -175,6 +179,111 @@ def mark_recently_accessed(self, server_name: Optional[str], media_id: str) -> N
...
+
+        if media.get("media_length") is not None:
+            raise SynapseError(
+                409,
+                "Media ID already has content",
+                errcode=Codes.CANNOT_OVERWRITE_MEDIA,
+            )
+
+        now = self.clock.time_msec()
+        if media.get("created_ts", now) < now - self.unused_expiration_time:
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/media/media_repository.py:256]: @sumnerevans Any thoughts on this?

**Comment by clokep** [comment]
Looks like I mistakenly set this for review again -- please shout when you want me to take another look!

File: synapse/rest/media/media_repository_resource.py [hunk 17]
```diff
@@ -80,6 +81,7 @@ def __init__(self, hs: "HomeServer"):
         super().__init__()
         media_repo = hs.get_media_repository()
 
+        self.putChild(b"create", CreateResource(hs, media_repo))
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/rest/media/media_repository_resource.py:84]: #16419 should make it possible to customize the registration here.

File: synapse/media/_base.py [hunk 18]
```diff
@@ -83,6 +83,53 @@
...
 ]
 
+# Default timeout_ms for download and thumbnail requests
+DEFAULT_MAX_TIMEOUT_MS = 20_000
+
+# Maximum allowed timeout_ms for download and thumbnail requests
+MAXIMUM_ALLOWED_MAX_TIMEOUT_MS = 60_000
+
+
+def parse_media_id(request: Request) -> Tuple[str, str, Optional[str]]:
```
Comment (clokep) [inline synapse/media/_base.py:93]: This coming back seems like a bad rebase?

Comment (sumnerevans) [reply synapse/media/_base.py:93]: I don't see any mention of this being a bad change, but this PR has so many comments...

Comment (clokep) [reply synapse/media/_base.py:93]: I think https://github.com/matrix-org/synapse/commit/b3126a95f4788834d450befe733106282ff35136 will fix it.

File: synapse/media/media_repository.py [hunk 19]
```diff
@@ -183,6 +187,120 @@ def mark_recently_accessed(self, server_name: Optional[str], media_id: str) -> N
...
+            media_type=media_type,
+            upload_name=upload_name,
+            media_length=content_length,
+            user_id=auth_user,
+        )
+
+        try:
+            await self._generate_thumbnails(None, media_id, media_id, media_type)
+        except Exception as e:
+logger.info("Failed to generate thumbnails: %s", e)
```
Comment (clokep) [inline synapse/media/media_repository.py:302]:
```suggestion
        try:
            await self._generate_thumbnails(None, media_id, media_id, media_type)
        except Exception as e:
            logger.info("Failed to generate thumbnails: %s", e)
```

File: synapse/media/media_repository.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/media/media_repository.py:264]: I think this is still valid?

When can `created_ts` ever be missing from the `media` info returned?

File: flake.nix
Comment (clokep) [inline flake.nix:1]: I think the changes in this file are unrelated.

File: synapse/media/media_repository.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (sumnerevans) [reply synapse/media/media_repository.py:266]: yes, looks correct. I used `.get` everywhere since it used to be a dictionary, but that seems to have been fixed

File: synapse/media/media_repository.py [hunk 20]
```diff
@@ -297,27 +470,35 @@ async def get_remote_media(
...
             upload_name = name if name else media_info["upload_name"]
             await respond_with_responder(
                 request, responder, media_type, media_length, upload_name
             )
+        elif media_info:
+            # The media exists (but the responder doesn't) meaning that this file hasn't
+            # been uploaded yet.
+            self.respond_not_yet_uploaded(request)
         else:
             respond_404(request)
```
[the remark this answers is not in the document]
Comment (sumnerevans) [reply synapse/media/media_repository.py]: I'm not really certain how this should work. All of the remote media stuff is still a bit confusing to me.

What you said makes sense, though, so I think it should just 404 here.

37c79704d0 should fix

File: docs/usage/configuration/config_documentation.md [hunk 21]
```diff
@@ -1792,6 +1805,27 @@ Example configuration:
...
+### `max_pending_media_uploads`
+
+How many *pending media uploads* can a given user have? A pending media upload
+is a created MXC URI that (a) is not expired (the `unused_expires_at` timestamp
+has not passed) and (b) the media has not yet been uploaded for. Defaults to 5.
+
+Example configuration:
+```yaml
+max_pending_media_uploads: 5
+```
```
[the remark this answers is not in the document]
Comment (sumnerevans) [reply docs/usage/configuration/config_documentation.md:1839]: This was from discussions in the MSC... I wanted to prove that implementations could provide multiple knobs for rate limiting.

If you are using async upload and uploading immediately, you could possibly be under this limit and still hit a rate limit (if the user sets it lower than the default or you have many small files or something).

[the remark this answers is not in the document]
Comment (sumnerevans) [reply docs/usage/configuration/config_documentation.md:1839]: That said, I am fine with adjusting what is provided in Synapse, because the MSC is merged, and we can provide whatever knobs make sense.

File: synapse/rest/media/thumbnail_resource.py [hunk 22]
```diff
@@ -142,8 +152,11 @@ async def _select_or_generate_local_thumbnail(
...
         desired_method: str,
         desired_type: str,
+        max_timeout_ms: int,
     ) -> None:
-        media_info = await self.store.get_local_media(media_id)
+        media_info = await self.media_repo.get_local_media_info(
+            request, media_id, max_timeout_ms
+        )
 
         if not media_info:
```
[the remark this answers is not in the document]
Comment (sumnerevans) [reply synapse/rest/media/thumbnail_resource.py:170]: I think that this is resolved, since it does return immediately if `media_info` is `None` now.

File: docs/usage/configuration/config_documentation.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (clokep) [reply docs/usage/configuration/config_documentation.md:1839]: > If you are using async upload and uploading immediately, you could possibly be under this limit and still hit a rate limit (if the user sets it lower than the default or you have many small files or something).

Oh, true! 👍

Overall -- I don't think this is a thing worth blocking the feature on though, so we can go with the current numbers.

File: synapse/rest/media/create_resource.py [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+        self._create_media_rate_limiter = Ratelimiter(
+            store=hs.get_datastores().main,
+            clock=self.clock,
+            cfg=hs.config.ratelimiting.rc_media_create,
+        )
+
+    async def _async_render_OPTIONS(self, request: SynapseRequest) -> None:
+        respond_with_json(request, 200, {}, send_cors=True)
+
+    async def _async_render_POST(self, request: SynapseRequest) -> None:
```
Comment (clokep) [inline synapse/rest/media/create_resource.py:53]: This is now a servlet so we can simplify a bit:

```suggestion
    async def on_POST(self, request: SynapseRequest) -> None:
```

File: synapse/storage/databases/main/media_repository.py [hunk 24]
```diff
@@ -425,6 +473,36 @@ async def mark_local_media_as_safe(self, media_id: str, safe: bool = True) -> No
...
+                    user_id.to_string(),
+                    self._clock.time_msec() - self.unused_expiration_time,
+                ),
+            )
+            row = txn.fetchone()
+            if not row:
+                return 0, 0
+            return row[0], (row[1] + self.unused_expiration_time if row[1] else 0)
+
+        return await self.db_pool.runInteraction("get_url_cache", get_pending_media_txn)
```
Comment (clokep) [inline synapse/storage/databases/main/media_repository.py:504]:
```suggestion
        def get_pending_media_txn(txn: LoggingTransaction) -> Tuple[int, int]:
            sql = """
            SELECT COUNT(*), MIN(created_ts)
            FROM local_media_repository
            WHERE user_id = ?
                AND created_ts > ?
                AND media_length IS NULL
            """
            txn.execute(
                sql,
                (
                    user_id.to_string(),
                    self._clock.time_msec() - self.unused_expiration_time,
                ),
            )
            row = txn.fetchone()
            if not row:
                return 0, 0
            return row[0], (row[1] + self.unused_expiration_time if row[1] else 0)

        return await self.db_pool.runInteraction("get_pending_media", get_pending_media_txn)
```

File: synapse/rest/media/upload_resource.py [hunk 25]
```diff
@@ -103,3 +114,53 @@ async def on_POST(self, request: SynapseRequest) -> None:
         respond_with_json(
             request, 200, {"content_uri": str(content_uri)}, send_cors=True
         )
+
+
+class AsyncUploadServlet(BaseUploadServlet):
+    PATTERNS = [
+        re.compile(
+            "/_matrix/media/v1/upload/(?P<server_name>[^/]*)/(?P<media_id>[^/]*)$"
```
Comment (clokep) [inline synapse/rest/media/upload_resource.py:122]: I was surprised to see that this is a [v3 endpoint according to the spec](https://spec.matrix.org/v1.8/client-server-api/#put_matrixmediav3uploadservernamemediaid)

```suggestion
            "/_matrix/media/v3/upload/(?P<server_name>[^/]*)/(?P<media_id>[^/]*)$"
```

I think it is worth keeping different patterns so that r0/v1 don't work though.

Comment (sumnerevans) [reply synapse/rest/media/upload_resource.py]: Yeah, the versioning scheme is stupid. Create is `/v1` while this is `/v3` presumably because `/v1/upload` already existed and could conflict? We should have just made both of them `/v3` so it's not confusing, but Travis disagrees and that's what ended up in the spec.

**Comment by clokep** [comment]
With the [last comment](https://github.com/matrix-org/synapse/pull/15503#discussion_r1391635653) and [updating the complement tests](https://github.com/matrix-org/complement/pull/641#issuecomment-1808924023) this is all passing for me locally again! 👍 

I'm pretty certain it is good to merge.

File: synapse/media/media_repository.py [hunk 26]
```diff
@@ -231,8 +347,71 @@ async def create_content(
...
+            request: The incoming request.
+            media_id: The media ID of the content. (This is the same as
+                the file_id for local content.)
+            max_timeout_ms: the maximum number of milliseconds to wait for the
+                media to be uploaded.
+
+        Returns:
+            Either the info dictionary for the given local media ID or
+            ``None``. If ``None``, then no further processing is necessary as
+            this function will send the necessary JSON response.
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/media/media_repository.py:375]: Just reviewed this again and this is the remaining comment (in addition to re-running black / fixing the formatting).

I'll then kick off complement CI again and I think we'll be good to go. 🎉

**Review by clokep — APPROVED** [review]
Should be good if the test run works. 🎉

File: synapse/storage/databases/main/media_repository.py [hunk 27]
```diff
@@ -149,6 +150,8 @@ def __init__(
             self._drop_media_index_without_method,
         )
 
+        self.unused_expiration_time = hs.config.media.unused_expiration_time
```
Comment (clokep) [inline synapse/storage/databases/main/media_repository.py:153]: Looks like this doesn't always exist, which is pretty annoying:

```suggestion
        if hs.config.media.can_load_media_repo:
            self.unused_expiration_time = hs.config.media.unused_expiration_time
        else:
            self.unused_expiration_time = None
```

File: synapse/storage/databases/main/media_repository.py [hunk 28]
```diff
@@ -425,6 +473,38 @@ async def mark_local_media_as_safe(self, media_id: str, safe: bool = True) -> No
...
+                AND created_ts > ?
+                AND media_length IS NULL
+            """
+            txn.execute(
+                sql,
+                (
+                    user_id.to_string(),
+                    self._clock.time_msec() - self.unused_expiration_time,
+                ),
+            )
```
Comment (clokep) [inline synapse/storage/databases/main/media_repository.py:498]:
```suggestion
            assert self.unused_expiration_time is not None
            txn.execute(
                sql,
                (
                    user_id.to_string(),
                    self._clock.time_msec() - self.unused_expiration_time,
                ),
            )
```

File: synapse/storage/databases/main/media_repository.py [hunk 29]
```diff
@@ -149,6 +150,11 @@ def __init__(
             self._drop_media_index_without_method,
         )
 
+        if hs.config.media.can_load_media_repo:
+            self.unused_expiration_time = hs.config.media.unused_expiration_time
+        else:
+            self.unused_expiration_time = None
```
Comment (clokep) [inline synapse/storage/databases/main/media_repository.py:156]: (I tested it this time. 😄 )

```suggestion
        if hs.config.media.can_load_media_repo:
            self.unused_expiration_time: Optional[
                int
            ] = hs.config.media.unused_expiration_time
        else:
            self.unused_expiration_time = None
```

**Comment by clokep** [comment]
Complement with workers was failing, hopefully the latest change fixes it: https://github.com/matrix-org/synapse/actions/runs/6877944714

**Comment by clokep** [comment]
Tests passed minus a known flake (https://github.com/matrix-org/sytest/issues/1201). 🎉  Thanks for working through this with me! 🚀
